src/Storages/StorageInMemoryMetadata.cpp

```diff
@@ -81,6 +90,63 @@ void StorageInMemoryMetadata::setComment(const String & comment_)
     comment = comment_;
 }
 
+void StorageInMemoryMetadata::setDefiner(const ASTSQLSecurity & sql_security)
+{
+    if (sql_security.definer)
+        definer = sql_security.definer->toString();
+
+    sql_security_type = sql_security.type;
+}
+
+UUID StorageInMemoryMetadata::getDefinerID(DB::ContextPtr context) const
+{
+    if (!definer)
+    {
+        if (const auto definer_id = context->getUserID())
+            return *definer_id;
+
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "No user in context for sub query execution.");
+    }
+
+    const auto & access_control = context->getAccessControl();
+    return access_control.getID<User>(*definer);
+}
+
+ContextMutablePtr StorageInMemoryMetadata::getSQLSecurityOverriddenContext(ContextPtr context) const
+{
+    if (!sql_security_type.has_value())
+        return Context::createCopy(context);
+
+    if (sql_security_type == ASTSQLSecurity::Type::INVOKER)
+        return Context::createCopy(context);
+
+    auto new_context = Context::createCopy(context->getGlobalContext());
+    new_context->applySettingsChanges(context->getSettingsRef().changes());
```

Comment:
Here `applySettingsChanges()`, `setQueryContext()` both look excessive and wrong.

Reply:
About `applySettingsChanges()` - actually I think it's useful here but let's call it after `setUser()` and clamping:

```
auto settings_changes = context->getSettingsRef().changes();

new_context->setUser(getDefinerID(context));
new_context->clampToSettingsConstraints(settings_changes, SettingSource::QUERY);
new_context->applySettingsChanges(settings_changes);
```

Reply:
`setQueryContext()` looks wrong, we must not copy the current query context (which is BTW often equal to just `context`) to the new context.

Reply:
No, both lines are required. Firstly, we need to copy all settings from the old context because we patch it during the query execution, and without those patched values, the execution can fail. 
`setQueryContext()` is also required because some steps in execution heavily rely on it. I debugged it **a lot** and came to the conclusion that it's easier to keep it as is.

Reply:
I actually remembered what was the problem with using `new_context` as a QueryContext. The issue was with context lifetime. The definer context exists only during the read/write view stage, and then it expires. So operations, which take multiple steps after reading (e.g., joins) will try to work with an expired context. 
Here is an example when I tried to use `new_context` as a query context: https://s3.amazonaws.com/clickhouse-test-reports/54901/6c59bbf82d5af9b79f2276616da4f0540050445b/fast_test.html

The good side is that we don't actually care what is inside our query context. We never check any permission outside the main context, so everything should work OK.